Login page and auth

andykwan1124 opened the pull request:
1. Any user trying to access the CMS without logging in will be redirected to the login page.
2. Users attempting to log in with an email that is not a @nwplus.io email will also be redirected, with an error message showing on the login page
3. Logged in users will first be taken to landing page
4. If logged in users try to access the login page they will also be taken to landing
5. Any logged in users clicking the sign out button will be pushed back to the login page and will need to relogin
6. Firebase Persistence is set to 'session', so if a user logs in to CMS in 1 tab, but opens the CMS in another tab, the user will need to login again in the new tab. Users don't need to re-enter the password to their gmail account though, they just need to click the 'login' button and they will be taken to landing page. I can't seem to force Google to require password input. Do we want to do that? Any suggestions are appreciated. 

After logging in, the current user object is available in the Auth context. You can access it by importing and calling useAuth()

Just a reminder that amplify link is available below if you want to test out some stuff. You can also review the backend auth in the firebase security rules and cloud functions.

Ciuffi commented on pages/index.js:
```diff
@@ -1,27 +1,72 @@
+  const checkAddAdminClaim = async user => {
+    const token = await user.getIdTokenResult();
+    if (!token.claims.hasOwnProperty('admin')) {
+      await setAdmin()
+      await firebase.auth().currentUser.getIdToken(true)
+    }
```
What is happening here?

Ciuffi commented on pages/index.js:
```diff
@@ -1,27 +1,72 @@
+    try {
+      await firebase.auth().signInWithPopup(provider)
+      const user = firebase.auth().currentUser
+      if (user && /.+@nwplus\.io$/.test(user.email)) {
+        router.push('/landing')
+        checkAddAdminClaim(user)
```
we should probably check the admin claims and use that as the source of truth for logging in, rather than the email.

Ciuffi commented on pages/landing.js:
```diff
@@ -1,3 +1,26 @@
-import React from 'react';
```
Is the landing page still needed?

Ciuffi commented on utility/auth.js:
```diff
@@ -0,0 +1,41 @@
+
+    return (
+        <AuthContext.Provider value={{ isAuthenticated: !!user, user }}>
+            {isLoading ? null : children}
```
Maybe add some loading text so its clear whats going on.

Ciuffi commented on utility/auth.js:
```diff
@@ -0,0 +1,41 @@
+    
+    useEffect(() => {
+        const unsubscribe = firebase.auth().onAuthStateChanged(async user => {
+            if (user && /.+@nwplus\.io$/.test(user.email)) {
+                setUser(user)
+                if (router.pathname=='/') await router.push('/landing')
+            } else {
+                await router.push('/')
+            }
+            setIsLoading(false)
```
again we should check for admin claims.

andykwan1124 commented:
1. Auth now solely relies on admin claim in user token instead of email
2. Updated design
3. Added a loading screen and status text on login page during authorization process

The background on login page is a png instead of svg because there are around 1k dots on it, and each of them being an svg might be a huge hit on performance. I used next-optimized-images to optimize the png but lighthouse still doesn't like it. Any suggestions are appreciated.

Also fyi the PR preview url for amplify seems to be deleted after around a week. Pushing new commits doesn't activate a new url.

Ciuffi commented:
@andykwan1124 good catch on the amplify thing, I was wondering what was going on. @kozr any idea on how to fix it?

Ciuffi commented on pages/index.js:
```diff
@@ -1,27 +1,109 @@
-        <CardContent>Example usage of card component</CardContent>
-      </Card>
+      <BackgroundContainer>
+          <LeftContainer>
```
Why the left container? I think it looks fine, just wondering if you got it from design or something. If not, could you post a picture of it and tag design?

andykwan1124 replied:
nah I didnt get it from design. I used that as the flex container for everything on login page other than the background. Should I still ping design and what should I ask them?

Ciuffi commented on utility/auth.js:
```diff
@@ -0,0 +1,72 @@
+
+    useEffect(() => {
+        const setPersistence = async () => {
+            await firebase.auth().setPersistence(firebase.auth.Auth.Persistence.SESSION)
```
can we use LOCAL persistence instead? This makes it so the user doesn't have to re authenticate every time they open the cms.

```suggestion
            await firebase.auth().setPersistence(firebase.auth.Auth.Persistence.LOCAL)
```

andykwan1124 replied:
ya for sure! I think firebase auth persistence is defaulted to local for web apps, so I'm just gonna remove that effect hook.

Ciuffi reviewed (approved):
two small comments but other than that, this looks and works great!

Ciuffi commented on .eslintrc.js:
```diff
@@ -1,23 +1,23 @@
+      es6: true,
+    },
+    plugins: ['react'],
+    extends: ['eslint:recommended', 'plugin:react/recommended'],
```
i would change this back to using the nextjs plugin instead. the react recommended eslint properties don't really work with nextjs that well and you'll never pass linting.

andykwan1124 commented:
testing amplify

kozr commented:
@Ciuffi Andy and I found that it's not an issue with it expiring, maybe someone clicked something on Amplify and stopped it from working. We'll try to set it up again.

kozr commented:
All evidence points to someone messing with amplify but failed to fix it. We will find the culprit and file a lawsuit and hold them responsible.

Ciuffi commented:
@kozr hmm I messed with amplify because I noticed it stopped working, but clearly i set it up wrong, could you redo it?

kozr commented:
if no one takes responsibility, we will fire the devs one by one starting with @andykwan1124

andykwan1124 commented:
@Ciuffi I reconnected the CMS repo to Amplify, still the old PRs weren't given a preview url even after closing and reopening them. New PRs have the urls tho (for now). Should I just open a new PR for this branch?

Ciuffi commented:
sure, just make my requested changes then open a new pr
